File: web.py
from threading import Thread
from typing import Optional, Any

from flask import Flask, render_template, request

from conf import current_conf, GLOBAL_CONFIG_SECTION_NAME, CONF_KEY_ENABLED_WEBSITES
from conf_meta import ConfMeta
from tool_logging import logger

app = Flask(__name__)
app.logger = logger


@app.route('/')
def index():
    return render_template('index.html', config_options=current_conf.export())


def check_single_config(lst: list, cfg_meta: ConfMeta) -> (Any, bool):
    from conf_meta import FIELD_TYPE_STR, FIELD_TYPE_STR_LIST, FIELD_TYPE_S_SELECTION, FIELD_TYPE_M_SELECTION, \
        FIELD_TYPE_BOOL
    if cfg_meta.typ == FIELD_TYPE_STR:
        return lst[0], True
    elif cfg_meta.typ == FIELD_TYPE_STR_LIST:
        if lst[0] == "":
            return [], True
        items: list[str] = lst[0].split(",")
        new_items = [item.strip() for item in items if item.strip() != ""]
        return new_items, True
    elif cfg_meta.typ == FIELD_TYPE_S_SELECTION:
        if len(lst) == 0:
            return None, False
        return lst[0], True
    elif cfg_meta.typ == FIELD_TYPE_M_SELECTION:
        return lst, True
    elif cfg_meta.typ == FIELD_TYPE_BOOL:
        return lst[0] == "True", True
    return None, False


def is_value_empty(value) -> bool:
    return value is None or value == "" or value == []


@app.route('/save_config', methods=["POST"])
def save_config():
    if request.method == "POST":
        current_export = current_conf.export()
        meta = current_export["meta"]

        # param check & get enabled agent list
        enabled_agents = request.form.getlist(CONF_KEY_ENABLED_WEBSITES)

        # param check
        for meta_item in meta:
            config_list: list[ConfMeta] = meta_item["configs"]
            if meta_item["section"] == GLOBAL_CONFIG_SECTION_NAME:
                for config in config_list:
                    value_in_form = request.form.getlist(config.config_key)
                    cfg_v, ok = check_single_config(value_in_form, config)
                    if ok:
                        if is_value_empty(cfg_v) and config.required:
                            return {"success": False, "message": f'Config "{config.config_name}" is required.'}
                    else:
                        return {"success": False, "message": f'Config "{config.config_name}" is not valid value.'}
            else:
                agent_name = meta_item["section"]
                for config in config_list:
                    form_key = agent_name + "__" + config.config_key
                    value_in_form = request.form.getlist(form_key)
                    cfg_v, ok = check_single_config(value_in_form, config)
                    if ok:
                        if config.required and agent_name in enabled_agents and is_value_empty(cfg_v):
                            return {"success": False,
                                    "message": f'Config "{config.config_name}" of "{agent_name}" is required.'}
                    else:
                        return {"success": False, "message": f'Config "{config.config_name}" is not valid value.'}

        for meta_item in meta:
            config_list: list[ConfMeta] = meta_item["configs"]
            if meta_item["section"] == GLOBAL_CONFIG_SECTION_NAME:
                for config in config_list:
                    value_in_form = request.form.getlist(config.config_key)
                    cfg_v, _ = check_single_config(value_in_form, config)
                    current_conf.dct[config.config_key] = cfg_v
            else:
                agent_name = meta_item["section"]
                for config in config_list:
                    form_key = agent_name + "__" + config.config_key
                    value_in_form = request.form.getlist(form_key)
                    cfg_v, _ = check_single_config(value_in_form, config)
                    current_conf.set_agent_conf(agent_name, config.config_key, cfg_v)
        current_conf.write_disk()
        # reflect new configuration
        return {"success": True}


def web_interface():
    app.run(debug=True, use_reloader=False, host='0.0.0.0')


def run_web_interface():
    t = Thread(target=web_interface, daemon=True)
    t.start()

File: web.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/gorilla/mux"
	"html/template"
	"io"
	"net/http"
	"reflect"
	"slices"
	"strings"
)

func runHTTPServer() {
	router := mux.NewRouter()
	router.HandleFunc("/", indexHandler).Methods("GET")
	router.HandleFunc("/save_config", saveConfigHandler).Methods("POST")
	router.HandleFunc("/otp", otpGetHandler).Methods("GET")
	router.HandleFunc("/otp", otpPostHandler).Methods("POST")
	http.ListenAndServe("0.0.0.0:5000", router)
}

func indexHandler(w http.ResponseWriter, r *http.Request) {
	exportedConf, err := currentConf.Export()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	tmpl := template.New("index.html")
	tmpl, err = tmpl.ParseFiles("template/index.html")
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	err = tmpl.Execute(w, exportedConf)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
}

func saveConfigHandler(w http.ResponseWriter, r *http.Request) {
	err := r.ParseForm()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	// get conf meta
	confExport, err := currentConf.Export()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	// param check
	for _, confSection := range confExport.ConfigSections {
		for _, config := range confSection.Configs {
			formValues := r.Form[confSection.Section+"__"+config.ConfigKey]
			isRequired := (confSection.Section == GlobalConfigSectionName && config.Required) ||
				(slices.Contains(r.Form[GlobalConfigSectionName+"__"+ConfKeyEnabledWebsites], confSection.Section) && config.Required)
			if isRequired && len(formValues) == 0 {
				json.NewEncoder(w).Encode(map[string]interface{}{"success": false,
					"message": fmt.Sprintf(`Config "%s" is required`, config.ConfigName)})
				return
			}
			if config.Type == FieldTypeString && len(formValues) > 0 {
				value := formValues[0]
				if isRequired && value == "" {
					json.NewEncoder(w).Encode(map[string]interface{}{"success": false,
						"message": fmt.Sprintf(`Config "%s" under section "%s" is required`, config.ConfigName, confSection.DisplayName)})
					return
				}
			} else if config.Type == FieldTypeBool && len(formValues) > 0 {
				value := formValues[0]
				if value != TrueLiteral && value != FalseLiteral {
					json.NewEncoder(w).Encode(map[string]interface{}{"success": false,
						"message": fmt.Sprintf(`Invalid value for bool option "%s"`, config.ConfigName)})
					return
				}
			}
		}
	}

	// set values
	for _, confSection := range confExport.ConfigSections {
		for _, config := range confSection.Configs {
			formValues := r.Form[confSection.Section+"__"+config.ConfigKey]
			if len(formValues) == 0 {
				continue
			}
			var structToChange reflect.Value
			if confSection.Section == GlobalConfigSectionName {
				// is global option
				structToChange = reflect.ValueOf(currentConf)
			} else {
				// is agent option
				if currentConf.AgentConfs[confSection.Section] == nil {
					currentConf.AgentConfs[confSection.Section] = &AgentConf{}
				}
				structToChange = reflect.ValueOf(currentConf.AgentConfs[confSection.Section])
			}

			var err error
			if config.Type == FieldTypeString {
				value := formValues[0]
				err = SetValueByJSONTag(structToChange, config.ConfigKey, value)
			} else if config.Type == FieldTypeStringList {
				value := strings.Split(formValues[0], ",")
				if len(value) == 1 && value[0] == "" {
					value = []string{}
				}
				err = SetValueByJSONTag(structToChange, config.ConfigKey, value)
			} else if config.Type == FieldTypeCheckbox {
				value := formValues
				err = SetValueByJSONTag(structToChange, config.ConfigKey, value)
			} else if config.Type == FieldTypeRadio {
				value := formValues[0]
				err = SetValueByJSONTag(structToChange, config.ConfigKey, value)
			} else if config.Type == FieldTypeBool {
				value := formValues[0]
				if value == "True" {
					err = SetValueByJSONTag(structToChange, config.ConfigKey, true)
				} else {
					err = SetValueByJSONTag(structToChange, config.ConfigKey, false)
				}
			} else {
				json.NewEncoder(w).Encode(map[string]interface{}{"success": false,
					"message": fmt.Sprintf("unknown field type %s", config.Type)})
				return
			}
			if err != nil {
				json.NewEncoder(w).Encode(map[string]interface{}{"success": false,
					"message": fmt.Sprintf("set value for %s__%s failed: %s", confSection.Section, config.ConfigKey, err.Error())})
				return
			}
		}
	}

	err = currentConf.WriteDisk()
	if err != nil {
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false,
			"message": fmt.Sprintf("write config to disk failed: %s", err.Error())})
		return
	}

	json.NewEncoder(w).Encode(map[string]interface{}{"success": true})
}

func otpGetHandler(w http.ResponseWriter, r *http.Request) {
	select {
	case reqAgent := <-otpRequestChan:
		json.NewEncoder(w).Encode(map[string]interface{}{"prompt": fmt.Sprintf("%s requires OTP. Please check your mailbox and input OTP here:", reqAgent)})
	default:
		json.NewEncoder(w).Encode(map[string]interface{}{"prompt": nil})
	}
}

func otpPostHandler(w http.ResponseWriter, r *http.Request) {
	type Payload struct {
		Input string `json:"input"`
	}

	body, _ := io.ReadAll(r.Body)
	defer r.Body.Close()

	var p Payload
	err := json.Unmarshal(body, &p)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	select {
	case otpChan <- p.Input:
		json.NewEncoder(w).Encode(map[string]interface{}{"success": true})
	default:
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "message": "No website requires OTP"})
	}
}
